### PR TITLE
Use hashed rekord type for tlog upload

### PR DESCRIPTION
**Pull request by asraa**
Signed-off-by: Asra Ali <asraa@google.com>

#### Summary

* Use a hashed rekord type for tlog upload to allow for signing large blobs with keyless flow.

**THIS IS A NOT A DRAFT ANYMORE BECAUSE...**
1. Prod rekor needs to be deployed with the new type **(DONE)**
2. The bigger issue is that if someone is using `COSIGN_EXPERIMENTAL` and uploading signed entries, let's say, with a KMS provider that uses a specific hash algorithm, then defaulting to the SHA256 I hard-coded in `rekorEntry` will be incorrect and will fail signature verification. The hash algorithm needs to be plumbed through, so I imagine that SignerVerifier would have to expose a new method to it's interface providing the hash function that it uses by default (doesn't look like cosign will ever use a `WithCryptoSignerOpts(otherHashFunc)` on its `SignMessage`s
- to get around this I guess I could specifically only have hashedRekord type being used for fulcio generated certs that we know will be using sha256? that'll at least help keyless sign-blobs. but i would rather not solve just a tiny part of the problem

#### Ticket Link

Fixes https://github.com/sigstore/rekor/issues/481

#### Release Note

```release-note

```

### CODE REVIEW

**Comment by dlorenc** [comment]
> 2\. The bigger issue is that if someone is using `COSIGN_EXPERIMENTAL` and uploading signed entries, let's say, with a KMS provider that uses a specific hash algorithm, then defaulting to the SHA256 I hard-coded in `rekorEntry` will be incorrect and will fail signature verification. The hash algorithm needs to be plumbed through, so I imagine that SignerVerifier would have to expose a new method to it's interface providing the hash function that it uses by default (doesn't look like cosign will ever use a `WithCryptoSignerOpts(otherHashFunc)` on its `SignMessage`s

The spec pretty much requires SHA256 right now. People can technically use other ones in a KMS, but I'd be fine fixing that later, maybe built on #1071

**Comment by asraa** [comment]
> The spec pretty much requires SHA256 right now. 

Do you mean the spec for keyless signing? Or more generally. I'm just worried about someone doing KMS w/ sha512 + rekor upload

**Comment by dlorenc** [comment]
> Do you mean the spec for keyless signing? Or more generally. I'm just worried about someone doing KMS w/ sha512 + rekor upload

The spec for all of cosign: https://github.com/sigstore/cosign/blob/main/specs/SIGNATURE_SPEC.md

**Comment by asraa** [comment]
> The spec for all of cosign: https://github.com/sigstore/cosign/blob/main/specs/SIGNATURE_SPEC.md

doh right, thanks! yeah i suppose working out the issues for sign-blobs with KMS can come later

File: go.mod [hunk 1]
```diff
@@ -55,3 +55,5 @@ require (
 	k8s.io/klog/v2 v2.20.0 // indirect
 	k8s.io/utils v0.0.0-20210930125809-cb0fa318a74b // indirect
 )
+
+replace github.com/sigstore/rekor => github.com/sigstore/rekor v0.3.1-0.20211119021052-57b3fef494e0
```
Comment (dlorenc) [inline go.mod:59]: we can probably do this with a normal bump instead of the replace here.

**Comment by dlorenc** [comment]
We should get this one in before v1.4.0 as well!

File: go.mod (hunk 1 above)
Comment (asraa) [reply go.mod:59]: I think we'll need a new release here in order to include hashedrekord type!

Comment (dlorenc) [reply go.mod:59]: Ah I think you can do something like: `go get github.com/sigstore/rekor@cae63bd4e389b56e352c7ffebbb86d1c1437409b` to pin to an unreleased commit.

Comment (asraa) [reply go.mod:59]: Argh thank you! sorry misinterpreted what you meant by normal bump!

**Comment by asraa** [comment]
QQ about this: while this change happens, should cosign look for both the rekord proposed entry and, if it doesn't exist, the hashedrekord proposed entry? i think it's okay, because two sigs are always gonna be unique because of randomization, so one or the other should exist, not both. (I suppose it's possible for someone to hand craft a rekord and hashedrekord with the same exact signature, but then it shouldn't matter unless we accidentally serve a different form of the public key, e.g. the x509 cert vs the extracted public key that misses some client verification)

**Comment by dlorenc** [comment]
> QQ about this: while this change happens, should cosign look for both the rekord proposed entry and, if it doesn't exist, the hashedrekord proposed entry? i think it's okay, because two sigs are always gonna be unique because of randomization, so one or the other should exist, not both. (I suppose it's possible for someone to hand craft a rekord and hashedrekord with the same exact signature, but then it shouldn't matter unless we accidentally serve a different form of the public key, e.g. the x509 cert vs the extracted public key that misses some client verification)

Just using the new format seems fine, most lookups go through the offline bundle verification anyway.

**Comment by asraa** [comment]
Done! Updated the offline bundle verification to unmarshal either or. Tested both container/blob verification

https://gist.github.com/asraa/c2ef0bd3aacb7073b1e0736c0f9fd557

File: pkg/cosign/verify.go [hunk 2]
```diff
@@ -469,20 +471,32 @@ func bundleHash(bundleBody, signature string) (string, string, error) {
 		return *intotoObj.Content.Hash.Algorithm, *intotoObj.Content.Hash.Value, nil
 	}
 
-	err = json.Unmarshal(bodyDecoded, &rekord)
+	if err := json.Unmarshal(bodyDecoded, &rekord); err == nil {
```
Comment (dlorenc) [inline pkg/cosign/verify.go:474]: did you double check this actually fails? Sometimes unmarshal will succeed even if it's the wrong type. We might need to get a bit more clever here.

Comment (asraa) [reply pkg/cosign/verify.go:474]: yep! can confirm that for a hashed rekord signature, it actually travels the second path and doesn't unmarshal to a rekord without an error

Comment (asraa) [reply pkg/cosign/verify.go:474]: i just checked how it errors: errors out on "invalid kind value" (rekor checks to make sure that the type we want to unmarshal to is the kind in the JSON)

Comment (dlorenc) [reply pkg/cosign/verify.go:474]: Awesome!

**Review by dlorenc — APPROVED** [review]

**Pull request merged**